apps/nextjs/src/middleware.ts

```diff
@@ -1,19 +1,86 @@
 import * as Sentry from "@sentry/nextjs";
-import { authMiddleware } from "middlewares/auth.middleware";
-import { addCspHeaders } from "middlewares/csp";
 import { NextMiddlewareResult } from "next/dist/server/web/types";
-import { NextMiddleware, NextResponse } from "next/server";
+import {
+  NextFetchEvent,
+  NextMiddleware,
+  NextRequest,
+  NextResponse,
+} from "next/server";
 
 import { sentryCleanup } from "./lib/sentry/sentryCleanup";
+import { authMiddleware } from "./middlewares/auth.middleware";
+import { addCspHeaders } from "./middlewares/csp";
+
+export type ErrorWithPotentialCause = Error & {
+  cause?: unknown;
+};
+
+export async function handleError(
+  error: unknown,
+  request: NextRequest,
+  event: NextFetchEvent,
+): Promise<Response> {
+  const requestInfo = {
+    url: request.url,
+    method: request.method,
+    headers: Object.fromEntries(request.headers),
+    cookies: Object.fromEntries(request.cookies),
+    geo: request.geo,
+    ip: request.ip,
+    nextUrl: {
+      pathname: request.nextUrl.pathname,
+      search: request.nextUrl.search,
+    },
+  };
+
+  const eventInfo = {
+    sourcePage: event.sourcePage,
+  };
+
+  let bodyText = "";
+  try {
+    const clonedRequest = request.clone();
+    bodyText = await clonedRequest.text();
+  } catch (bodyError) {
+    console.error("Failed to read request body", bodyError);
+    bodyText = "Failed to read body";
+  }
+
+  // Cast error to our custom type
+  const errorWithCause = error as ErrorWithPotentialCause;
+
+  if (
+    error instanceof SyntaxError ||
+    errorWithCause.cause instanceof SyntaxError
+  ) {
+    const syntaxError =
+      error instanceof SyntaxError
+        ? error
+        : (errorWithCause.cause as SyntaxError);
+
+    console.warn({
+      event: "middleware.syntaxError",
+      error: syntaxError.message,
+      url: requestInfo.url,
+      request: requestInfo,
+    });
 
-function handleError(error: unknown, extra: Record<string, unknown>): Response {
-  if (error instanceof SyntaxError) {
-    console.error("Handled SyntaxError in nextMiddleware", error);
     return NextResponse.json({ error: "Bad Request" }, { status: 400 });
   }
 
-  const wrappedError = new Error("Error in nextMiddleware", { cause: error });
-  Sentry.captureException(wrappedError, { extra });
+  const wrappedError =
+    error instanceof Error
+      ? error
+      : new Error("Error in nextMiddleware", { cause: error });
+
+  Sentry.captureException(wrappedError, {
+    extra: {
+      requestInfo,
+      eventInfo,
+      bodyText,
+      errorType: "OtherError",
+    },
+  });
```

Comment:
This looks smart. Let's extract it somewhere and co-locate with the test

Reply:
Agreed - I'll extract it to a utility function